cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+	Properties RuntimeProperties
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIdRunningYet(config *Config, runsAddr string, runId string) (bool, error) {
+	s, err := getProgramRunStatusByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return false, err
+	}
+
+	if s == "RUNNING" {
+		return true, nil
+	}
+	if !INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+		return false, fmt.Errorf("Program not Running or Initializing, in state: %s", s)
+	}
+	return false, nil
+
+}
+
+func getProgramRunStatusByFauxId(config *Config, runsAddr string, runId string) (s string, err error) {
+	r, err := getRunByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return
+	}
+
+	return r.Status, nil
+}
+
+// TODO optimization: we call this function often when we could probably get the real runid once and cache it.
+// This would avoid redoing this loop everytime to get the same result. probably inconsequential unless there are many runs of this program
+func getRunByFauxId(config *Config, runsAddr string, runId string) (r Run, err error) {
+	req, err := http.NewRequest(http.MethodGet, runsAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []Run
+	var args RuntimeArgs
+
+	err = json.Unmarshal(b, &runs)
+	if err != nil {
+		return
+	}
+
+	for _, r := range runs {
+		// Unescaping runtime Args which are stored as an escaped JSON string.
+		unquotedArgs, _ := strconv.Unquote(string(r.Properties.RuntimeArgs))
+		log.Println(fmt.Sprintf("%s", unquotedArgs))
+		b := []byte(unquotedArgs)
+		if err := json.Unmarshal(b, &args); err != nil {
+			// This happens when a run does not contain the special TF con
+			log.Println("failed to parse unquotedArgs json")
+		}
+		log.Println(fmt.Sprintf("found terraform faux run id: %s", args.FauxRunId))
+		log.Println(fmt.Sprintf("status: %s", r.Status))
+		if runId == args.FauxRunId {
+			return r, nil
+		}
+	}
+	err = fmt.Errorf("no run found with faux runid: %s", runId)
+	return
+}
+
+func stopProgramRun(config *Config, stopAddr string) error {
+	req, err := http.NewRequest(http.MethodPost, stopAddr, nil)
+	if err != nil {
+		return err
+	}
+	_, err = httpCall(config.httpClient, req)
+	return err
+}
+
+func resourceStreamingProgramRunDelete(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	runsAddr := urlJoin(addr, "/runs")
+	r, err := getRunByFauxId(config, runsAddr, d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	stopAddr := urlJoin(runsAddr, r.RunId, "/stop")
+
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		s, err := getProgramRunStatusByFauxId(config, runsAddr, d.Id())
+
+		if s == "RUNNING" || INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+			err = stopProgramRun(config, stopAddr)
```

Comment:
What happens if we always call stopProgramRun (regardless of status)?

Reply:
400 Bad Request: program run foo is in bar state